internal/pkg/reconcilers/bucket/capa/capa_reconciler.go

```diff
@@ -0,0 +1,189 @@
+	logger := log.FromContext(ctx)
+
+	originalBucket := bucket.DeepCopy()
+	// If the Bucket doesn't have our finalizer, add it.
+	if controllerutil.AddFinalizer(bucket, v1alpha1.BucketFinalizer) {
+		// Register the finalizer immediately to avoid orphaning AWS resources on delete
+		if err := r.Client.Patch(ctx, bucket, client.MergeFrom(originalBucket)); err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	logger.Info("Checking if bucket exists")
+	exists, err := s3Service.BucketExists(ctx, bucket)
+	if err != nil {
+		logger.Info(fmt.Sprintf("Either you don't have access to bucket %v or another error occurred. "+
+			"Here's what happened: %v", bucket.Spec.Name, err))
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	if exists {
+		logger.Info(fmt.Sprintf("Bucket %v exists and you already own it.", bucket.Spec.Name))
+	} else {
+		logger.Info(fmt.Sprintf("Bucket %v is available, creating", bucket.Spec.Name))
+		err = s3Service.CreateBucket(ctx, bucket)
+		if err != nil {
+			logger.Error(err, "Bucket could not be created")
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	if bucket.Spec.Acl != nil {
+		logger.Info("Setting up bucket acl")
+		err = s3Service.SetBucketACL(ctx, bucket)
+		if err != nil {
+			logger.Error(err, "Bucket ACL could not be set")
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	logger.Info("Setting up lifecycle rule")
+	// If expiration is not set, we remove all lifecycle rules
+	err = s3Service.SetLifecycleRules(ctx, bucket)
+	if err != nil {
+		logger.Error(err, "Bucket lifecycle configuration could not be set")
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	logger.Info("Setting up bucket tags")
+	err = s3Service.SetTags(ctx, bucket)
+	if err != nil {
+		logger.Error(err, "Tags could not be applied")
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	logger.Info("Bucket ready")
+	originalBucket = bucket.DeepCopy()
+	bucket.Status.BucketID = bucket.Spec.Name
+	bucket.Status.BucketReady = true
+	if err := r.Client.Status().Patch(ctx, bucket, client.MergeFrom(originalBucket)); err != nil {
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileDelete deletes the s3 bucket.
+func (r CAPABucketReconciler) reconcileDelete(ctx context.Context, s3Service *s3.Service, bucket *v1alpha1.Bucket) error {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Checking if bucket exists")
+	_, err := s3Service.BucketExists(ctx, bucket)
+	if err == nil {
+		logger.Info("Bucket exists, deleting")
+		err = s3Service.DeleteBucket(ctx, bucket)
+		if err != nil {
+			logger.Error(err, "Bucket could not be deleted")
+			return errors.WithStack(err)
+		}
+	}
+
+	logger.Info("Bucket deleted")
+	// Bucket is deleted so remove the finalizer.
+	controllerutil.RemoveFinalizer(bucket, v1alpha1.BucketFinalizer)
+	return r.Client.Update(ctx, bucket)
```

Comment:
Better to use `patch` than `update` to avoid conflicts.

Reply:
Fixed